toe/src/urbit/app/toe.hoon

```diff
@@ -0,0 +1,921 @@
+      ++  turno   ^-(toe-turno [player pos])
+      ++  out     (outcome-check player board.room)
+      --
+  |%
+  ++  outcome-check
+    |=  [per=player =board-game]
+    ^-  outcome
+    ?:  (wins per board-game)
+      `%wins
+    ?.  (ties board-game)
+      ~
+    `%tie
+  ::
+  ++  wins
+    |=  [per=player =board-game]
+    ^-  ?
+    %+  lien  winning-rows
+    |=  a=(list spot)
+    ^-  ?
+    %+  levy  a
+    |=  b=spot
+    =/  p=(unit player)  (~(get by board-game) b)
+    ?~(p %.n =(stone.per stone.u.p))
+  ::
+  ++  ties
+    |=(=board-game =(~(wyt in board-game) 9))
+  ::
+  ++  winning-rows
+    ^-  (list (list spot))
+    :~  ~[[%1 %1] [%2 %1] [%3 %1]]
+        ~[[%1 %2] [%2 %2] [%3 %2]]
+        ~[[%1 %3] [%2 %3] [%3 %3]]
+        ~[[%1 %1] [%1 %2] [%1 %3]]
+        ~[[%2 %1] [%2 %2] [%2 %3]]
+        ~[[%3 %1] [%3 %2] [%3 %3]]
+        ~[[%1 %3] [%2 %2] [%3 %1]]
+        ~[[%1 %1] [%2 %2] [%3 %3]]
+    ==
+  --
+::  +connect: handles subscriptions and updates
+::
+++  connect
+  |=  pax=path
+  ^-  (quip card _state)
+  ~&  who-connect+pax
+  |^  ?+    pax  ~|([%peer-toe-strange pax] !!)
+          [%sole *]           sole
+          [%toetile ~]        tile
+          [%room ^]           (room i.t.pax)
+          [%http-response *]  [~ state]
+      ==
+  ::  +tile: connects to the frontend
+  ::
+  ++  tile
+    ^-  (quip card _state)
+    ~&  tiled+src.bol
+    :_  state
+    [%give %fact ~ %json !>(*json)]~
+  ::  +sole: innitializes the console's state
+  ::
+  ++  sole
+    ^-  (quip card _state)
+    ~&  sole+[me src.bol]
+    ?.  (team:title me src.bol)  ~|([%strange-sole src.bol] !!)
+    :_  state(state.cli *sole-share)
+    =,  co:view
+    [(send mor+~[clear welcome wopr grid shall-we (prompt choose)])]~
+  ::  +room: receives a subscription to connecto to a room
+  ::
+  ++  room
+    |=  id=@t
+    ^-  (quip card _state)
+    =,  room-core
+    ?~  rooms  [~ state]
+    =/  ze=@p  (slav %p id)
+    ~&  [ze ship:current]
+    ?.  =(ze ship:current)  [~ state]
+    ?:  =(%wait game)
+      (join ze)
+    ?.  =(%play game)  [~ state]
+    (start ze)
+  --
+::  +view
+::  TODO: please make me into a %view app!
+::
+++  view
+  |%
+  ::  TODO: move other view updates here
+  ::
+  ++  updates
+    |%
+    ++  restart
+      ^-  (list card)
+      :~  (send:co mor+all-effects:co)
+          (send:fe [%status s+'null']~)
+      ==
+    ::
+    ++  receive
+      |=  guest=@p
+      =/  room-name=tape  (cite:title guest)
+      ?^  rooms
+        (notify guest)
+      :~  (send:fe ~[[%message s+(crip room-name)] [%status s+'confirm']])
+          =,  co
+          (send (prompt "{confirm}{room-name}? (Y/N) | "))
+      ==
+    ::
+    ++  confirm
+      =/  next=@p  guest:next:room-core
+      :~  (confirm:updates:fe next)
+          (confirm:updates:co next)
+      ==
+    ::
+    ++  playing
+      |=  move=toe-turno
+      ^-  (list card)
+      =/  [ze=@p who=@p]   =,  current:room-core
+        [ship who]
```

Comment:
Perhaps more clear:

```hoon
=/  [ze=@p who=@p]
  [ship who]:current:room-core
```

Reply:
I want to remember that I had done this but got a nest-fail? maybe not, I like it more so I'll try :)